### ISSUE TITLE
rules_docker rule for haskell_binary

### ISSUE THREAD

**Issue by jamesthompson** [tweag/rules_haskell#1216]
It's currently not clear what the best practices are for haskell binaries and their associated portability, e.g. deployment within docker containers.
 
I was wondering if anybody might be able to comment on whether anybody has successfully deployed a rules_haskell built haskell_binary in a docker container (and if so - how), and whether or not any thought has been given about how to (if at all possible) build a rule for rules_docker?

A good alternative to rules_docker would just be a simple example showing how to deploy a haskell_binary in a docker container in this project.

Any guidance at all would be greatly appreciated from the folks who work on this incredible project. Thanks in advance for your help!

**Comment by jamesthompson** [comment]
Following up on this. I have had some luck patching generated executable loaders using patchelf. haskell_binary is currently generated executables being linked with the nixpgs_cc_configure stuff against a nix/store path, with a different ld- name than what you might find on debian / ubuntu distroless for example `ld-2.27.so` or `ld-linux-x86-64.so.2` for example.

**Comment by regnat** [comment]
Hi @jamesthompson. We do deploy  `rules_haskell`-built docker images in some project.
A way to do this is that we use Nix to build a docker image with all the runtime dependencies of the executable, and use this image as the `base` in `cc_image`.

To build the image with Nix, you can use [dockerTools.buildLayeredImage](https://nixos.org/nixpkgs/manual/#ssec-pkgs-dockerTools-buildLayeredImage).

The total setup could look like:

- base_image.nix

    ```nix
    with import <nixpkgs> {};
    
    # Build the base image.
    # The output of this derivation will be a docker archive in the same format as
    # the output of `docker save` that we can feed to
    # [container_load](https://github.com/bazelbuild/rules_docker#container_load)
    let myImage =dockerTools.buildLayeredImage {
        name = "base-image-unwrapped";
        contents = [ glibc ]; # Add anything your executable might need at runtime that's provided by Nix
      };
    in
    # rules_nixpkgs require the nix output to be a directory,
    # so just create one in which we put the image we've just created
    runCommand "base-image" {} ''
    mkdir -p $out
    ln -s ${myImage} $out/image
    ''
    ```

- WORKSPACE

    ```bzl
    #### Import the nix-generated docker image.
    
    # First import the raw tarball with rules-nixpkgs
    nixpkgs_package(
      name = "raw-base-image",
      nix_file = "//:base_image.nix",
      build_file_content = """
    package(default_visibility = [ "//visibility:public" ]),
    exports_file(["image"])
      """,
    )
    
    # Then import it in `rules_docker` so that
    # we can use it as the `base` for another image
    container_load(
      name = "base-image",
      file = "@raw-base-image//:image",
    )
    ```

- BUILD.bazel

    ```bzl
    # The binary we want to wrap
    haskell_binary(
      name = "my-binary",
      ...
    )
    
    # Build a docker image out of this,
    # Using the previously built image as the base
    # (instead of distroless)
    cc_image(
      name = "my-binary-image",
      binary = "my-binary",
      base = "@base_image//image",
    )
    ```


------

A possible alternative is to use [clodl](https://github.com/tweag/clodl) to get a pseudo-static executable, although that's more fragile in my experience

**Comment by mboes** [comment]
Still, we should have a more convenient way of doing this, and better story for deploying Haskell binaries in general. See #379.

**Comment by jamesthompson** [comment]
@regnat Thank you very much indeed for this rapid help!

I've just been trying this on a linux vm against `rules_docker 0.13.0` which is compatible with rules_haskell master as far as I can tell.

I've set everything up exactly as how you describe here, but I'm seeing a docker load / import error 

```
ERROR: An error occurred during the fetch of repository 'base-image':
   Importing from tarball failed (status 1): 2020/01/20 13:09:13 Running the Image Loader to load the image tarball...
2020/01/20 13:09:13 Failed to load tarball into OCI format: failed to read docker image tarball from "/home/james/.cache/bazel/_bazel_james/7c0697dca6976ef12976f6658697d628/external/base-image/image": archive/tar: invalid tar header
```

Is there something else I could be missing? Do you know which version of rules_docker you or somebody else might have used with rules_haskell to make this work?

**Comment by jamesthompson** [comment]
Update: I just needed to gunzip the tarball created in the runCommand  "base-image" step and output that. 

Now I can't get the output bazel-built docker layer to boot the binary. Not sure if I need to use container_image instead. I'll give that a go. Thanks very much again for your help.

**Comment by jamesthompson** [comment]
Importing all the runfiles config is apparently taken care of by a script written by rules_docker. I successfully managed to call bazel run //target.. and that booted the container after configuring the docker toolchain optional config as described https://github.com/bazelbuild/rules_docker/blob/master/README.md#setup.

For others who may be interested, this is how I got rules_docker 0.13.0 to pull in the haskell base image...

```
runCommand "base-image" {} ''
  mkdir -p $out
  gunzip -c ${myImage} > $out/image
''
```

Running bazel run then loads the image into the docker daemon. I haven't tried the container push step yet, that's next for me. But I'm going to close out this issue. Thanks to all again for all your help and this great project.
